src/gui/render/graphics_shaders/shader_program.hpp

```diff
@@ -0,0 +1,387 @@
+// -*- lsst-c++ -*-
+/*
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ */
+
+/**
+ * @file shader_program.hpp
+ *
+ * @author @AlemSnyder
+ *
+ * @brief Defines shader program classes
+ *
+ * @ingroup GUI  RENDER  GRAPHICS_SHADERS
+ *
+ */
+
+#pragma once
+
+#include "../../handler.hpp"
+#include "../array_buffer/gpu_data.hpp"
+#include "gui_render_types.hpp"
+#include "logging.hpp"
+#include "types.hpp"
+#include "uniform.hpp"
+
+#include <algorithm>
+#include <functional>
+#include <memory>
+#include <utility>
+#include <vector>
+
+namespace gui {
+
+namespace shader {
+
+inline void
+log_uniforms(
+    std::set<std::pair<std::string, std::string>> want_uniforms,
+    std::set<std::pair<std::string, std::string>> has_uniforms
+) {
+#if DEBUG()
+
+    std::set<std::pair<std::string, std::string>> uniforms_needed;
+    std::set<std::pair<std::string, std::string>> uniforms_not_needed;
+
+    std::set_difference(
+        want_uniforms.begin(), want_uniforms.end(), has_uniforms.begin(),
+        has_uniforms.end(), std::inserter(uniforms_needed, uniforms_needed.begin())
+    );
+
+    std::set_difference(
+        has_uniforms.begin(), has_uniforms.end(), want_uniforms.begin(),
+        want_uniforms.end(),
+        std::inserter(uniforms_not_needed, uniforms_not_needed.begin())
+    );
+
+    if (uniforms_needed.size() > 0)
+        LOG_WARNING(logging::opengl_logger, "Needed uniforms {}", uniforms_needed);
+    if (uniforms_not_needed.size() > 0)
+        LOG_DEBUG(
+            logging::opengl_logger, "Not needed uniforms {}", uniforms_not_needed
+        );
+
+#endif
+}
+
+class ShaderProgram_Standard : virtual public render_to::FrameBuffer {
+ private:
+    Program& opengl_program_;
+
+    const std::function<void()> setup_;
+
+    Uniforms uniforms_;
+
+ public:
+    // Ya I know this looks bad, but data_ is basically a parameter
+    std::vector<std::shared_ptr<array_buffer::GPUData>> data;
```

Comment:
just make it private and expose it with a `data()` method returning a reference

Reply:
no

Reply:
Why

Reply:
Why bother making it privite when there is no meaningfull difference?

Reply:
Because everyone expects it to be?

`std::vector` does it with the `data()` method.